cpp/src/gis/point_in_polygon.cu

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2019 BlazingDB, Inc.
+ *     Copyright 2019 Christian Cordova Estrada <christianc@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "gis.hpp"
+#include "point_in_polygon.hpp"
+
+#include <bitmask/BitMask.cuh>
+#include "bitmask/bitmask_ops.hpp"
+#include "utilities/type_dispatcher.hpp"
+
+#include <type_traits>
+
+namespace {
+
+    /** 
+ * @brief Compute the orientation of p3 from two others points p1 and p2
+ *
+ * @param[in] p1_x: Longitude of the first point p1
+ * @param[in] p1_y: Latitude of the first point p1
+ * @param[in] p2_x: Longitude of the second point p2
+ * @param[in] p2_y: Latitude of the second point p2
+ * @param[in] p3_x: Longitude of the third point p3
+ * @param[in] p3_y: Latitude of the third point p3
+ *
+ * @returns positive if it's clockwise, negative if is counter-clockwise and 0 if is colinear
+ */
+template <typename T>
+__device__ T orientation(const T p1_x, const T p1_y, const T p2_x, const T p2_y, const T p3_x, const T p3_y)
+{
+    return ((p2_y - p1_y) * (p3_x - p2_x) - (p2_x - p1_x) * (p3_y - p2_y));
+}
+
+/** 
+ *  @brief Determine whether or not coordinates (query points) are completely inside a static polygon
+ *
+ * @param[in] poly_lats: Pointer to latitudes of a polygon
+ * @param[in] poly_lons: Pointer to longitudes of a polygon
+ * @param[in] point_lats: Pointer to latitudes of many query points
+ * @param[in] point_lons: Pointer to longitudes of many query points
+ * @param[in] poly_size: Size of polygon
+ * @param[in] point_size: Total number of query points
+ * @param[out] point_is_in_polygon: Pointer indicating if the i-th query point is inside or not
+ *
+ * @returns
+ */
+ template <typename T>
+ __global__ void point_in_polygon_kernel(const T* poly_lats,
```

Comment:
I assume that these pointers will never point to overlapping buffers? If so, add `__restrict__`. 
```suggestion
 __global__ void point_in_polygon_kernel(const T* const __restrict__ poly_lats,
```

Reply:
Better! For read-only and not overlapping.